contracts/example/SnapshotGate/SnapshotGate.sol

```diff
@@ -0,0 +1,335 @@
+ * - Create Offers to be gated on the protocol
+ *   - the qty available for an offer should match the supply of its corresponding snapshot token
+ * - Create Groups on the protocol which
+ *   - wrap their corresponding offers
+ *   - have a condition that
+ *     - expects a specific token (ERC721)
+ *     - uses this contract address as the token address
+ *     - uses the appropriate snapshot token id for the gated offer
+ *     - has maxCommits setting that matches the supply of its corresponding snapshot token
+ */
+contract SnapshotGate is BosonTypes, Ownable, ERC721 {
+    // Event emitted when the snapshot is appended to
+    event SnapshotAppended(Holder[] holders);
+
+    // Event emitted when the snapshot is frozen
+    event SnapshotFrozen();
+
+    // Event emitted when a buyer commits via this gate
+    event SnapshotTokenCommitted(address indexed buyer, uint256 indexed offerId, uint256 indexed tokenId);
+
+    // Token holders and their amounts
+    struct Holder {
+        uint256 tokenId;
+        uint256 amount;
+        address owner;
+    }
+
+    // Transactional state
+    enum TransactionStatus {
+        NotInTransaction,
+        InTransaction
+    }
+
+    // Details of in-flight transaction
+    struct TransactionDetails {
+        address buyer;
+        uint256 tokenId;
+    }
+
+    // Transaction status
+    TransactionStatus private txStatus;
+
+    // When txStatus is InTransaction,
+    // This is the transaction details
+    TransactionDetails private txDetails;
+
+    // Address of the Boson Protocol
+    address protocol;
+
+    // Id of the seller operating the snapshot
+    uint256 sellerId;
+
+    // Is the snapshot frozen
+    bool snapshotFrozen;
+
+    // Track holders in snapshot
+    // token id => owner => total owned
+    mapping(uint256 => mapping(address => uint256)) snapshot;
+
+    // Track committed tokens
+    // token id => owner => total committed
+    mapping(uint256 => mapping(address => uint256)) committed;
+
+    // Modifier to check whether a gate transaction is in progress
+    modifier statusCheck() {
+        txStatus = TransactionStatus.InTransaction;
+        _;
+        txStatus = TransactionStatus.NotInTransaction;
+    }
+
+    // Add safeTransferFrom to IERC20
+    using SafeERC20 for IERC20;
+
+    // Constructor
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _protocol,
+        uint256 _sellerId
+    ) ERC721(_name, _symbol) {
+        protocol = _protocol;
+        sellerId = _sellerId;
+        txStatus = TransactionStatus.NotInTransaction;
+    }
+
+    /**
+     * Append a batch of holders to the snapshot
+     *
+     * Reverts if:
+     * - Caller is not contract owner
+     * - Snapshot is frozen
+     *
+     * @param _holders an array of Holder structs
+     */
+    function appendToSnapshot(Holder[] memory _holders) public onlyOwner {
```

Comment:
`_holders` can be calldata, not memory.

Reply:
Yeah, it was weird, when @dennisfurrer and I were pairing on this to start, I told him to use calldata, but it complained that it had to be memory or storage. Will revisit.

Reply:
Fixed in next commit